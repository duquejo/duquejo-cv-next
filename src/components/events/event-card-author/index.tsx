export * from './linked-author';
export * from './static-author';
