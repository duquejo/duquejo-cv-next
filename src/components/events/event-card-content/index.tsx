export * from './push-event';
export * from './watch-event';
export * from './blog-post-event';
