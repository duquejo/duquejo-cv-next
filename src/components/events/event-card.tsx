import type { Event } from '@/interfaces';
import Link from 'next/link';
import Image from 'next/image';
import { Eye, GitBranch, Github, GitMerge, GitPullRequest } from 'lucide-react';
import {
  Card,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from '@/components/ui/card';
import { ReactNode } from 'react';
import { Badge } from '@/components/ui/badge';
import { cn } from '@/lib';

export const EventCard = ({ created_at, payload, actor, repo, type }: Event) => {
  const toLocaleDateString = (dateString: string) => {
    return new Date(dateString).toLocaleString();
  };

  const toIcon = (iconString: string): ReactNode => {
    const commonCss = 'inline bg-primary rounded-full transition-colors p-1';
    const commonProps = {
      size: 30,
      'data-testid': 'event-action',
    };
    switch (iconString) {
      case 'PullRequestEvent':
        return <GitPullRequest className={cn(commonCss, 'bg-yellow-400')} {...commonProps} />;
      case 'CreateEvent':
        return <GitBranch className={cn(commonCss, 'bg-teal-400')} {...commonProps} />;
      case 'PushEvent':
        return <GitMerge className={cn(commonCss, 'bg-red-400')} {...commonProps} />;
      case 'WatchEvent':
      case 'PullRequestReviewEvent':
        return <Eye className={cn(commonCss, 'bg-cyan-300')} {...commonProps} />;
      default:
        return <GitMerge className={cn(commonCss, 'bg-purple-400')} {...commonProps} />;
    }
  };

  return (
    <Card role="listitem" className="lg:max-w-xs">
      <CardHeader className="pb-3">
        <CardTitle className="flex items-center text-xs justify-between">
          <time className="text-xs font-extrabold">{toLocaleDateString(created_at)}</time>
          {payload.ref && <Badge variant="secondary">{payload.ref}</Badge>}
        </CardTitle>
        <CardDescription>
          <Link
            href={`https://github.com/${repo.name}`}
            className="flex items-center space-x-2"
            rel="noopener noreferrer nofollow"
            target="_blank"
          >
            <Image
              src={actor.avatar_url}
              alt={actor.display_login}
              className="max-h-7 rounded-full"
              width="28"
              height="28"
            />
            <span className="font-semibold underline underline-offset-2 decoration-1 truncate text-xs">
              {repo.name}
            </span>
          </Link>
        </CardDescription>
      </CardHeader>
      <CardContent className="text-xs leading-relaxed text-left break-words">
        {payload.description && <strong className="font-bold">{payload.description}</strong>}
        {payload.commits &&
          payload.commits.length > 0 &&
          payload.commits.map(
            (commit) =>
              !['dependabot[bot]'].includes(commit.author.name) && (
                <div key={commit.sha}>
                  <p className="inline text-muted-foreground">{commit.message}</p>
                  {type === 'PushEvent' && (
                    <Link
                      href={`${commit.url.replace('https://api.github.com/repos', 'https://github.com')}`}
                      className="inline font-semibold"
                      target="_blank"
                      rel="noopener noreferrer nofollow"
                      title="Link"
                    >
                      <Github className="ml-2 inline bg-primary rounded-full p-1 h-5 w-5 text-background transition-colors hover:bg-primary-foreground hover:text-foreground" />
                    </Link>
                  )}
                </div>
              ),
          )}
      </CardContent>
      <CardFooter className="flex justify-between">
        <span className="text-background">{toIcon(type)}</span>
        <span className="text-muted-foreground">{type}</span>
        {payload.ref_type && <span className="capitalize">({payload.ref_type})</span>}
      </CardFooter>
    </Card>
  );
};
