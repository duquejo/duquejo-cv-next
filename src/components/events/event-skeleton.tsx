import { Skeleton } from '@/components/ui/skeleton';

interface Props {
  rounds?: number;
}

export const EventSkeleton = ({ rounds = 2 }: Props) => {
  return new Array(rounds).fill(0).map((_, i) => (
    <div key={i} className="flex flex-col space-y-3">
      <div className="flex items-center justify-between gap-x-3">
        <Skeleton className="flex-1 h-8 w-full" />
        <Skeleton className="h-8 w-[100px]" />
      </div>
      <Skeleton className="h-[125px] w-full" />
      <div className="space-y-3">
        <Skeleton className="h-4 w-full" />
        <Skeleton className="h-4 w-full" />
      </div>
      <div className="flex items-center space-x-4">
        <Skeleton className="h-12 w-12 rounded-full" />
        <div className="space-y-3">
          <Skeleton className="h-4 w-[250px]" />
          <Skeleton className="h-4 w-[200px]" />
        </div>
      </div>
    </div>
  ));
};
