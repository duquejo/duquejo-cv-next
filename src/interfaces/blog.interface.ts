import { ReactNode } from 'react';

export interface BlogPost {
  slug: string;
  title: string;
  excerpt: string;
  category: string;
  tags: string[];
  publishDate: string;
  readingTime: string;
  coverImage: string;
}

export interface BlogPostResult {
  Post: () => ReactNode;
  metadata: BlogPost;
}
