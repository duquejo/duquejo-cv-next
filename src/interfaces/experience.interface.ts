export interface ExperienceType {
  start_date: string;
  end_date?: string;
  enterprise: string;
  isRecent: boolean;
  project: string;
  role: string;
  resume?: string[];
  additional_info?: string[];
  screenshot?: string;
  url?: string;
}
