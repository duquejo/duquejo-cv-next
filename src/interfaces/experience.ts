export interface ExperienceType {
  start_date: string;
  end_date: string;
  enterprise: string;
  isRecent: boolean;
  project: string;
  role: string;
  resume?: string[];
  additional_info?: string;
}

export type Sizes = 'small' | 'large' | 'medium';
