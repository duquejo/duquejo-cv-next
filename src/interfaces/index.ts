export * from './experience.interface';
export * from './menu-items.interface';
export * from './main-languages.interface';
export * from './services.interface';
export * from './knowledge.interface';
