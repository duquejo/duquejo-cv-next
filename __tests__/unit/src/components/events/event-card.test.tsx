import { cleanup, render, screen } from '@testing-library/react';
import { EventCard } from '@/components/events/event-card';
import { type Event, EventType } from '@/interfaces';

describe('<EventCard /> tests', () => {
  const args: Event = {
    id: '123',
    type: EventType.PushEvent,
    actor: {
      id: 123,
      url: 'https://foo.bar',
      avatar_url: 'https://foo.bar',
      login: 'johndoe',
      display_login: 'johndoe',
      gravatar_id: '123',
    },
    repo: {
      url: 'https://foo.bar',
      id: 213,
      name: 'johndoe',
    },
    payload: {
      commits: [],
    },
    public: false,
    created_at: '2024-01-23T16:16:53.477Z',
  };

  it('should match the snapshot', () => {
    const { container } = render(<EventCard {...args} />);

    expect(container).toMatchSnapshot();
  });

  it('should render the optional fields if they are given', () => {
    const completeArgs: Event = {
      ...args,
      payload: {
        ref: 'main', // Optional
        ref_type: 'PushEvent', // Optional
        description: 'Commit title',
        commits: [
          {
            // Optional
            author: {
              name: 'johndoe',
              email: 'johndoe@email.com',
            },
            url: 'https://api.github.com/repos/johndoe',
            sha: '123',
            message: 'Commit detail',
            distinct: false,
          },
        ],
      },
    };
    const { container } = render(<EventCard {...completeArgs} />);

    expect(container).toMatchSnapshot();

    expect(screen.getByText(completeArgs.payload.ref!)).toBeInTheDocument();
    expect(screen.getByRole('link', { name: 'Link' })).toHaveAttribute(
      'href',
      'https://github.com/johndoe',
    );
    expect(screen.getByRole('link', { name: 'Link' })).toHaveAttribute(
      'rel',
      'noopener noreferrer nofollow',
    );
    expect(screen.getByText(completeArgs.payload.description!)).toBeInTheDocument();
    expect(screen.getByText(completeArgs.payload.ref_type!)).toBeInTheDocument();
  });

  it('should render the event icon conditionally', () => {
    const events = [
      {
        type: EventType.CreateEvent,
        expectedColor: 'bg-teal-400',
      },
      {
        type: EventType.PullRequestEvent,
        expectedColor: 'bg-yellow-400',
      },
      {
        type: EventType.PushEvent,
        expectedColor: 'bg-red-400',
      },
      {
        type: EventType.WatchEvent,
        expectedColor: 'bg-cyan-300',
      },
      {
        type: EventType.PullRequestReviewEvent,
        expectedColor: 'bg-cyan-300',
      },
      {
        type: 'OtherEventType',
        expectedColor: 'bg-purple-400',
      },
    ];

    events.forEach((event) => {
      const eventConfig: Event = {
        ...args,
        type: event.type,
      };
      render(<EventCard {...eventConfig} />);

      expect(screen.getByTestId('event-action')).toBeInTheDocument();
      expect(screen.getByTestId('event-action')).toHaveClass(event.expectedColor);

      cleanup();
    });
  });
});
