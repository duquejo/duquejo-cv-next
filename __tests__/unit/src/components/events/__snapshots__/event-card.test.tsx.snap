// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<EventCard /> tests > should match the snapshot 1`] = `
<div>
  <div
    class="rounded-xl border bg-card text-card-foreground shadow-sm lg:max-w-xs"
    role="listitem"
  >
    <div
      class="flex flex-col space-y-1.5 p-6 pb-3"
    >
      <div
        class="font-semibold tracking-tight flex items-center text-xs justify-between"
      >
        <time
          class="text-xs font-extrabold"
        >
          23/1/2024, 11:16:53
        </time>
      </div>
      <div
        class="text-sm text-muted-foreground"
      >
        <a
          class="flex items-center space-x-2"
          href="https://github.com/johndoe"
          rel="noopener noreferrer nofollow"
          target="_blank"
        >
          <img
            alt="johndoe"
            class="max-h-7 rounded-full"
            data-nimg="1"
            decoding="async"
            height="28"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Ffoo.bar&w=64&q=75"
            srcset="/_next/image?url=https%3A%2F%2Ffoo.bar&w=32&q=75 1x, /_next/image?url=https%3A%2F%2Ffoo.bar&w=64&q=75 2x"
            style="color: transparent;"
            width="28"
          />
          <span
            class="font-semibold underline underline-offset-2 decoration-1 truncate text-xs"
          >
            johndoe
          </span>
        </a>
      </div>
    </div>
    <div
      class="p-6 pt-0 text-xs leading-relaxed text-left break-words"
    />
    <div
      class="items-center p-6 pt-0 flex justify-between"
    >
      <span
        class="text-background"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-git-merge inline rounded-full transition-colors p-1 bg-red-400"
          data-testid="event-action"
          fill="none"
          height="30"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="18"
            cy="18"
            r="3"
          />
          <circle
            cx="6"
            cy="6"
            r="3"
          />
          <path
            d="M6 21V9a9 9 0 0 0 9 9"
          />
        </svg>
      </span>
      <span
        class="text-muted-foreground"
      >
        PushEvent
      </span>
    </div>
  </div>
</div>
`;

exports[`<EventCard /> tests > should render the optional fields if they are given 1`] = `
<div>
  <div
    class="rounded-xl border bg-card text-card-foreground shadow-sm lg:max-w-xs"
    role="listitem"
  >
    <div
      class="flex flex-col space-y-1.5 p-6 pb-3"
    >
      <div
        class="font-semibold tracking-tight flex items-center text-xs justify-between"
      >
        <time
          class="text-xs font-extrabold"
        >
          23/1/2024, 11:16:53
        </time>
        <div
          class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-hidden focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80"
        >
          main
        </div>
      </div>
      <div
        class="text-sm text-muted-foreground"
      >
        <a
          class="flex items-center space-x-2"
          href="https://github.com/johndoe"
          rel="noopener noreferrer nofollow"
          target="_blank"
        >
          <img
            alt="johndoe"
            class="max-h-7 rounded-full"
            data-nimg="1"
            decoding="async"
            height="28"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Ffoo.bar&w=64&q=75"
            srcset="/_next/image?url=https%3A%2F%2Ffoo.bar&w=32&q=75 1x, /_next/image?url=https%3A%2F%2Ffoo.bar&w=64&q=75 2x"
            style="color: transparent;"
            width="28"
          />
          <span
            class="font-semibold underline underline-offset-2 decoration-1 truncate text-xs"
          >
            johndoe
          </span>
        </a>
      </div>
    </div>
    <div
      class="p-6 pt-0 text-xs leading-relaxed text-left break-words"
    >
      <strong
        class="font-bold"
      >
        Commit title
      </strong>
      <div>
        <p
          class="inline text-muted-foreground"
        >
          Commit detail
        </p>
        <a
          class="inline font-semibold"
          href="https://github.com/johndoe"
          rel="noopener noreferrer nofollow"
          target="_blank"
          title="Link"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-github ml-2 inline bg-primary rounded-full p-1 h-5 w-5 text-background transition-colors hover:bg-primary-foreground hover:text-foreground"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 22v-4a4.8 4.8 0 0 0-1-3.5c3 0 6-2 6-5.5.08-1.25-.27-2.48-1-3.5.28-1.15.28-2.35 0-3.5 0 0-1 0-3 1.5-2.64-.5-5.36-.5-8 0C6 2 5 2 5 2c-.3 1.15-.3 2.35 0 3.5A5.403 5.403 0 0 0 4 9c0 3.5 3 5.5 6 5.5-.39.49-.68 1.05-.85 1.65-.17.6-.22 1.23-.15 1.85v4"
            />
            <path
              d="M9 18c-4.51 2-5-2-7-2"
            />
          </svg>
        </a>
      </div>
    </div>
    <div
      class="items-center p-6 pt-0 flex justify-between"
    >
      <span
        class="text-background"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-git-merge inline rounded-full transition-colors p-1 bg-red-400"
          data-testid="event-action"
          fill="none"
          height="30"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="18"
            cy="18"
            r="3"
          />
          <circle
            cx="6"
            cy="6"
            r="3"
          />
          <path
            d="M6 21V9a9 9 0 0 0 9 9"
          />
        </svg>
      </span>
      <span
        class="text-muted-foreground"
      >
        PushEvent
      </span>
      <span
        class="capitalize"
      >
        (
        PushEvent
        )
      </span>
    </div>
  </div>
</div>
`;
